### DISCUSSION TITLE
关于选择器查询结果数量的疑问

### DISCUSSION THREAD

**Discussion by adproqwq** [gkd-kit/gkd#569] [Q&A] [answered]
不知道在这里询问是否合适？

https://i.gkd.li/i/14730844
在该快照中，目标节点是[_id=264]的ViewGroup节点，
使用以下选择器：
```text
[name$="RecyclerView"] >2 * >n @[name$="ImageView" || name$="ViewGroup"][clickable=true] <<n [childCount<=5] < * >n [text*="广告"][visibleToUser=true]
```
却只选择到了[_id=259]的ImageView节点，目标节点经我推演，也应该符合条件（如果的确不符合条件的话请关闭当前discussion），但是没有被搜索到。
这是怎么回事？

**Comment by lisonge** [comment]
如果符合条件，你推演出来的选择路径是什么？

**Comment by adproqwq** [comment]
> 如果符合条件，你推演出来的选择路径是什么？

下面推演出来的选择器省略了部分修饰性属性，以及对`<<`的语义修改为：`>`的相反运算（即A >4 B 等价于 B <<4 A）
```text
RecyclerView >2 ViewGroup >3 @ViewGroup[_id=264] <<3 ViewGroup[childCount=5] < ViewGroup >4 [text="广告"]

RecyclerView >2 ViewGroup >3 @ViewGroup[_id=264] <<2 ViewGroup < ViewGroup[childCount=5] >3 [text="广告"]
```

**Comment by lisonge** [comment, marked as answer]
只从查询结果数量上看，数量为 1 是正确的

因为选择目标始终是 `[text*="广告"][visibleToUser=true]`

选择器是原理是遍历所有节点，然后检测这个节点是否满足你给出的  **此节点作为起点的路径约束**

如果满足就返回路径上的节点作为一个数组，`@` 只是决定选取这个数组的索引

尽管可能有多条子约束符合条件，但是只会选取最先得到的一个约束

比如 `@* > [_id=20]` 和 `@* >n [_id=20]` 没有区别

因为 `@* >n [_id=20]` 在 n=1 时是 `@* > [_id=20]` 此时已经找到符合条件的约束，自然不会再找 n=2,3,4... 之后的约束路径

**Reply by adproqwq** [reply]
懂了
